src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -0,0 +1,919 @@
+
+                if let Abi::RustIntrinsic | Abi::PlatformIntrinsic = self.tcx.fn_sig(def_id).abi() {
+                    assert!(!self.tcx.is_const_fn(def_id));
+
+                    if self.tcx.item_name(def_id) == sym::transmute {
+                        self.check_op(ops::Transmute);
+                        return;
+                    }
+
+                    // To preserve the current semantics, we return early, allowing all
+                    // intrinsics (except `transmute`) to pass unchecked to miri.
+                    //
+                    // FIXME: We should keep a whitelist of allowed intrinsics (or at least a
+                    // blacklist of unimplemented ones) and fail here instead.
+                    return;
+                }
+
+                if self.tcx.is_const_fn(def_id) {
+                    return;
+                }
+
+                if is_lang_panic_fn(self.tcx, def_id) {
+                    self.check_op(ops::Panic);
+                } else if let Some(feature) = self.tcx.is_unstable_const_fn(def_id) {
+                    // Exempt unstable const fns inside of macros with
+                    // `#[allow_internal_unstable]`.
+                    if !self.span.allows_unstable(feature) {
+                        self.check_op(ops::FnCallUnstable(def_id, feature));
+                    }
+                } else {
+                    self.check_op(ops::FnCallNonConst(def_id));
+                }
+
+            }
+
+            // Forbid all `Drop` terminators unless the place being dropped is a local with no
+            // projections that cannot be `NeedsDrop`.
+            | TerminatorKind::Drop { location: dropped_place, .. }
+            | TerminatorKind::DropAndReplace { location: dropped_place, .. }
+            => {
+                let mut err_span = self.span;
+
+                // Check to see if the type of this place can ever have a drop impl. If not, this
+                // `Drop` terminator is frivolous.
+                let ty_needs_drop = dropped_place
+                    .ty(self.body, self.tcx)
+                    .ty
+                    .needs_drop(self.tcx, self.param_env);
+
+                if !ty_needs_drop {
+                    return;
+                }
+
+                let needs_drop = if let Place {
+                    base: PlaceBase::Local(local),
+                    projection: box [],
+                } = *dropped_place {
+                    // Use the span where the local was declared as the span of the drop error.
+                    err_span = self.body.local_decls[local].source_info.span;
+                    self.qualifs.needs_drop.contains(local)
+                } else {
+                    true
+                };
+
+                if needs_drop {
+                    self.check_op_spanned(ops::LiveDrop, err_span);
+                }
+            }
+
+            _ => {}
+        }
+    }
+}
+
+/// All implementers of `NonConstOp`.
+pub mod ops {
```

Comment:
since this module makes up a third of the file, please split it into its own file